interfaces/mount/spec.go

```diff
@@ -158,33 +159,115 @@ func (spec *Specification) MountEntries() []osutil.MountEntry {
 	result = append(result, spec.overname...)
 	result = append(result, spec.layout...)
 	result = append(result, spec.general...)
-	unclashMountEntries(result)
-	return result
+	return unclashMountEntries(result)
 }
 
 // UserMountEntries returns a copy of the added user mount entries.
 func (spec *Specification) UserMountEntries() []osutil.MountEntry {
 	result := make([]osutil.MountEntry, len(spec.user))
 	copy(result, spec.user)
-	unclashMountEntries(result)
-	return result
+	return unclashMountEntries(result)
+}
+
+// Assuming that two mount entries have the same source, target and type, this
+// function computes the mount options that should be used when performing the
+// mount, so that the most permissive options are kept.
+// The following flags are considered (of course the operation is commutative):
+//   - "ro" + "rw" = "rw"
+//   - "bind" + "rbind" = "rbind
+func mergeOptions(options ...[]string) []string {
+	mergedOptions := make([]string, 0, len(options[0]))
+	foundWritableEntry := false
+	foundRBindEntry := false
+	firstEntryIsBindMount := false
+	for i, opts := range options {
+		isReadOnly := false
+		isRBind := false
+		for _, o := range opts {
+			switch o {
+			case "ro":
+				isReadOnly = true
+			case "rbind":
+				isRBind = true
+				fallthrough
+			case "bind":
+				if i == 0 {
+					firstEntryIsBindMount = true
+				}
+			case "rw", "async":
+				// these are default options for mount, do nothing
+			default:
+				// write all other options
+				if !strutil.ListContains(mergedOptions, o) {
+					mergedOptions = append(mergedOptions, o)
+				}
+			}
+		}
+		if !isReadOnly {
+			foundWritableEntry = true
+		}
+		if isRBind {
+			foundRBindEntry = true
+		}
+	}
+
+	if !foundWritableEntry {
+		mergedOptions = append(mergedOptions, "ro")
+	}
+
+	if firstEntryIsBindMount {
+		if foundRBindEntry {
+			mergedOptions = append(mergedOptions, "rbind")
+		} else {
+			mergedOptions = append(mergedOptions, "bind")
+		}
+	}
+
+	return mergedOptions
 }
 
 // unclashMountEntries renames mount points if they clash with other entries.
 //
 // Subsequent entries get suffixed with -2, -3, etc.
 // The initial entry is unaltered (and does not become -1).
-func unclashMountEntries(entries []osutil.MountEntry) {
-	count := make(map[string]int, len(entries))
+func unclashMountEntries(entries []osutil.MountEntry) []osutil.MountEntry {
+	result := make([]osutil.MountEntry, 0, len(entries))
+
+	type clashingEntry struct {
+		// Indexes of the entries in the `entries` array
+		Indexes []int
```

Comment:
we could also have:

FirstIndex int
Count int

instead as we never read anything but Indexes[0] and otherwise we should use its length?

Reply:
Makes sense, done!